stl/inc/ranges

```diff
@@ -9169,6 +9169,632 @@ namespace ranges {
+                return _Out < _Left || _Out < _Right;
+            } else {
+                using _UInt = _Make_unsigned_like_t<_Int>;
+                _Out        = static_cast<_Int>(static_cast<_UInt>(_Left) + static_cast<_UInt>(_Right));
+                return (_Left > 0 && _Right > 0 && _Out <= 0) || (_Left < 0 && _Right < 0 && _Out >= 0);
+            }
+        }
+    }
+
+    template <_Integer_like _Int>
+    _NODISCARD constexpr bool _Multiply_with_overflow_check(_Int _Left, _Int _Right, _Int& _Out) {
+#ifdef __clang__
+        if constexpr (integral<_Int>) {
+            return __builtin_mul_overflow(_Left, _Right, &_Out);
+        } else
+#endif // __clang__
+        {
+            if constexpr (!_Signed_integer_like<_Int>) {
+                _Out = _Left * _Right;
+                return _Left != 0 && _Right != 0 && (_Out < _Left || _Out < _Right);
+            } else {
+                // Based on llvm::MulOverflow
+                // See https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/MathExtras.h
+                using _UInt          = _Make_unsigned_like_t<_Int>;
+                const _UInt _ULeft   = _Left < 0 ? (0 - static_cast<_UInt>(_Left)) : static_cast<_UInt>(_Left);
+                const _UInt _URight  = _Right < 0 ? (0 - static_cast<_UInt>(_Right)) : static_cast<_UInt>(_Right);
+                const _UInt _UResult = _ULeft * _URight;
+
+                const bool _Negative = (_Left < 0) ^ (_Right < 0);
+                _Out                 = static_cast<_Int>(_Negative ? (0 - _UResult) : _UResult);
+                if (_ULeft == 0 || _URight == 0) {
+                    return false;
+                }
+
+                if (_Negative) {
+                    return _ULeft > (static_cast<_UInt>((numeric_limits<_Int>::max)()) + _UInt{1}) / _URight;
+                } else {
+                    return _ULeft > static_cast<_UInt>((numeric_limits<_Int>::max)()) / _URight;
+                }
+            }
+        }
+    }
+
+    template <bool _Const, class _First, class... _Rest>
+    concept _Cartesian_product_is_random_access = (random_access_range<_Maybe_const<_Const, _First>> && ...
+                                                   && (random_access_range<_Maybe_const<_Const, _Rest>>
+                                                       && sized_range<_Maybe_const<_Const, _Rest>>) );
+
+    template <class _Rng>
+    concept _Cartesian_product_common_arg = common_range<_Rng> || (sized_range<_Rng> && random_access_range<_Rng>);
+
+    template <bool _Const, class _First, class... _Rest>
+    concept _Cartesian_product_is_bidirectional = (bidirectional_range<_Maybe_const<_Const, _First>> && ...
+                                                   && (bidirectional_range<_Maybe_const<_Const, _Rest>>
+                                                       && _Cartesian_product_common_arg<_Maybe_const<_Const, _Rest>>) );
+
+    template <class _First, class... _Rest>

```

Comment:
Quick drive-by since I just suggested removing the unused template parameter pack from the Standard: remove `class... _Rest` here ;) There are only two uses to cleanup, and there's a minimal throughput benefit.